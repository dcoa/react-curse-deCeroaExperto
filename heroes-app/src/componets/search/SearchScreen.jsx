import { useNavigate, useLocation } from "react-router-dom";
import { getHeroesByTerm } from "../../helpers/getHeroesByTerm";
import { useForm } from "../../hooks/useForm"
import { HeroCard } from "../hero/HeroCard";
import queryString from 'query-string'
import { useMemo } from "react";


export const SearchScreen = () => {
    
    const navigate = useNavigate();
    const location = useLocation();
    // si no existe q es un string vacio por defecto
    const {q = ''} = queryString.parse(location.search);
    const { formValues, handleInputChange } = useForm({
        heroText: q
    });
    const { heroText } = formValues;
    const searchedHeroes = useMemo(() => getHeroesByTerm(q), [q]);
    const handleSearch = (e) => {
        e.preventDefault();
        navigate(`?q=${heroText}`)
    }
    return (
        <div>
            <div className="w-full border-b border-gray-300 mb-5 pb-5">
                <h1 className="text-2xl	text-slate-600">Search</h1>
            </div>
            <div>
                <h2 className="text-l text-slate-600 mb-3">Search a hero</h2>
                <form onSubmit={handleSearch} aria-label="search-form"
                className="flex flex-col md:flex-row"
                >
                    <input
                        type="text"
                        placeholder="Search...."
                        className=" h-10 focus:ring-2 focus:ring-blue-500 focus:outline-none appearance-none w-full text-sm leading-6 text-slate-900 placeholder-slate-400 rounded-md py-2 pl-10 ring-1 ring-slate-200 shadow-sm md:w-1/3 md:my-5"
                        name="heroText"
                        value={heroText}
                        autoComplete="off"
                        onChange={handleInputChange}
                    />
                    <button
                        type="submit"
                        className="h-10 px-6 w-full font-semibold rounded-md bg-black my-5 text-white md:w-1/3 md:ml-5"
                    >Submit</button>
                </form>
            </div>
            <div className="mt-5">
            <div className="w-full border-b border-gray-300 mb-5 pb-5">
            <h2 className="text-l	text-slate-600">Results</h2>
            </div>
            <ul className="mt-6 flex flex-wrap justify-around">

            { 
                (q === '') ? 
                <div>Start a new search</div>
                :
                (searchedHeroes.length === 0)
                ?
                <div> Not found a hero </div>

                :
                searchedHeroes.map( hero => (<HeroCard
                    key={hero.id}
                    {...hero} />))
            }
                        </ul>
            </div>
        </div>
    )
}