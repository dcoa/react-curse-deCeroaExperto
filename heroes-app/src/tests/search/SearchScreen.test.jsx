import { fireEvent, render, screen } from "@testing-library/react"
import { MemoryRouter, useNavigate } from "react-router-dom";
import { SearchScreen } from "../../componets/search/SearchScreen"

const mockNavigate = jest.fn();
jest.mock('react-router-dom', ()=>({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockNavigate,
}));

describe('SearchScreen', () => {
    beforeEach(() => jest.clearAllMocks());

    test('should show the default values', () => {
        const { container } = render((<MemoryRouter>
            <SearchScreen />
        </MemoryRouter>));

        expect(container).toMatchSnapshot();
    });

    test('should show Batman and the imput with the same queryString', () => {
        render((<MemoryRouter initialEntries={['/search?q=batman']}>
            <SearchScreen />
        </MemoryRouter>));
        const input = screen.getByRole('textbox');
        const image = screen.getByRole('img');
        expect(input.value).toBe('batman');
        expect(image.src).toContain('dc-batman.jpg');
    });

    test('should show a not found message if not exist the hero', () => {
        render((<MemoryRouter initialEntries={['/search?q=joker']}>
            <SearchScreen />
        </MemoryRouter>));
        const notFound = screen.getAllByText('Not found a hero');
        expect(notFound).toBeTruthy();
    });

    test('should show navigate to the hero page', () => {
        render((<MemoryRouter initialEntries={['/search']}>
            <SearchScreen />
        </MemoryRouter>));
        const input = screen.getByRole('textbox');
        const form = screen.getByRole('form');
        fireEvent.change(input, {target: {name: 'heroText', value: 'superman'}});
        fireEvent.submit(form);

        expect(mockNavigate).toHaveBeenCalledWith(`?q=superman`);
    });
})