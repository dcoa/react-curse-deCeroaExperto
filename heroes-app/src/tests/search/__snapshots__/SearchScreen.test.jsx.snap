// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchScreen should show the default values 1`] = `
<div>
  <div>
    <div
      class="w-full border-b border-gray-300 mb-5 pb-5"
    >
      <h1
        class="text-2xl	text-slate-600"
      >
        Search
      </h1>
    </div>
    <div>
      <h2
        class="text-l text-slate-600 mb-3"
      >
        Search a hero
      </h2>
      <form
        aria-label="search-form"
        class="flex flex-col md:flex-row"
      >
        <input
          autocomplete="off"
          class=" h-10 focus:ring-2 focus:ring-blue-500 focus:outline-none appearance-none w-full text-sm leading-6 text-slate-900 placeholder-slate-400 rounded-md py-2 pl-10 ring-1 ring-slate-200 shadow-sm md:w-1/3 md:my-5"
          name="heroText"
          placeholder="Search...."
          type="text"
          value=""
        />
        <button
          class="h-10 px-6 w-full font-semibold rounded-md bg-black my-5 text-white md:w-1/3 md:ml-5"
          type="submit"
        >
          Submit
        </button>
      </form>
    </div>
    <div
      class="mt-5"
    >
      <div
        class="w-full border-b border-gray-300 mb-5 pb-5"
      >
        <h2
          class="text-l	text-slate-600"
        >
          Results
        </h2>
      </div>
      <ul
        class="mt-6 flex flex-wrap justify-around"
      >
        <div>
          Start a new search
        </div>
      </ul>
    </div>
  </div>
</div>
`;
